Sources/UI/Device/Device.swift

```diff
@@ -0,0 +1,405 @@
+//
+//  Device.swift
+//  RoverUI
+//
+//  Created by Andrew Clunis on 2018-11-20.
+//  Copyright © 2018 Rover Labs Inc. All rights reserved.
+//
+
+import Foundation
+import os
+import UIKit
+import UserNotifications
+
+class Device {
+    
+    let userDefaults: UserDefaults
+    let jsonDecoder: JSONDecoder
+    let jsonEncoder: JSONEncoder
+    
+    let adSupportInfoProvider: AdSupportInfoProvider?
+    let bluetoothInfoProvider: BluetoothInfoProvider?
+    let telephonyInfoProvider: TelephonyInfoProvider?
+    let locationInfoProvider: LocationInfoProvider?
+    
+    let reachability = Reachability(hostname: "google.com")!
+    let userNotificationCenter = UNUserNotificationCenter.current()
+    
+    public init(
+        userDefaults: UserDefaults,
+        jsonDecoder: JSONDecoder,
+        jsonEncoder: JSONEncoder,
+        adSupportInfoProvider: AdSupportInfoProvider?,
+        bluetoothInfoProvider: BluetoothInfoProvider?,
+        telephonyInfoProvider: TelephonyInfoProvider?,
+        locationInfoProvider: LocationInfoProvider?
+    ) {
+        self.userDefaults = userDefaults
+        self.jsonDecoder = jsonDecoder
+        self.jsonEncoder = jsonEncoder
+        self.adSupportInfoProvider = adSupportInfoProvider
+        self.bluetoothInfoProvider = bluetoothInfoProvider
+        self.telephonyInfoProvider = telephonyInfoProvider
+        self.locationInfoProvider = locationInfoProvider
+    }
+    
+    public var snapshot: DeviceSnapshot {
+        return DeviceSnapshot(
+            advertisingIdentifier: advertisingIdentifier,
+            isBluetoothEnabled: isBluetoothEnabled,
+            localeLanguage: localeLanguage,
+            localeRegion: localeRegion,
+            localeScript: localeScript,
+            isLocationServicesEnabled: isLocationServicesEnabled,
+            location: location,
+            locationAuthorization: locationAuthorization,
+            notificationAuthorization: notificationAuthorization,
+            pushToken: pushToken,
+            isCellularEnabled: isCellularEnabled,
+            isWifiEnabled: isWifiEnabled,
+            appBadgeNumber: appBadgeNumber,
+            appBuild: appBuild,
+            appIdentifier: appIdentifier,
+            appVersion: appVersion,
+            buildEnvironment: buildEnvironment,
+            deviceIdentifier: deviceIdentifier,
+            deviceManufacturer: deviceManufacturer,
+            deviceModel: deviceModel,
+            deviceName: deviceName,
+            operatingSystemName: operatingSystemName,
+            operatingSystemVersion: operatingSystemVersion,
+            screenHeight: screenHeight,
+            screenWidth: screenWidth,
+            sdkVersion: sdkVersion,
+            carrierName: carrierName,
+            radio: radio,
+            isTestDevice: isTestDevice,
+            timeZone: timeZone,
+            userInfo: userInfo
+        )
+    }
+    
+    public private(set) var pushToken: DeviceSnapshot.PushToken? {
```

Comment:
This, along with all other computed properties, needs to be `open` as opposed to `public` so the developer can customize the implementation.